### PR TITLE
Adding Tags

### PR DESCRIPTION
**Pull request by melomys**
I added a tags attribute to Case and functionality in the form to choose tags and show statistics of them in the Metrics view

### CODE REVIEW

File: app/controllers/metrics_controller.rb [hunk 1]
```diff
@@ -2,18 +2,42 @@ class MetricsController < ApplicationController
...
+    tags_to_map = [["Health",["hospital", "health", "medical", "psychologist", "pregnant"]],
+                   ["Legal",["legal", "interview appointment", "interview prep", "family reunification", "first decision", "second decision"]],
+                   ["Clothes",["clothes", "freeshop", "free shop"]],
+                   ["Food", ["food"]]]
+    mapped_tags = []
+
+    tags_to_map.each do |tags|
+      if tags[1].any? {|k| description.downcase.include? k}
+        mapped_tags.push(tags[0])
+      end
```
Comment (NickPapacostas) [inline app/controllers/metrics_controller.rb:15]: A more idiomatic way to write this function is:

```
  def map_description(description)
    tags_with_keywords = {
      "Health" => ["hospital", "health", "medical", "psychologist", "pregnant"],
      "Legal" => ["legal", "interview appointment", "interview prep", "family reunification", "first decision", "second decision"],
      "Clothes" => ["clothes", "freeshop", "free shop"],
      "Food" => ["food"]
    }

    tags_with_keywords.select do |tag, kws|
      kws.any? {|kw| description.downcase.include?(kw) }
    end.map {|tag, kws| tag }
  end
```

File: app/controllers/metrics_controller.rb [hunk 2]
```diff
@@ -2,18 +2,42 @@ class MetricsController < ApplicationController
...
+    tags = Case.tag_options.map{|tag| [tag, []]}.to_h
+    cases.each do |c|
+      # if the case has no tags, try to read it out of the description
+      c_tags = c.tags.empty? ? map_description(c.description) : c.tags
+      c_tags.each do |tag|
+        unless tag == ""
+          tags[tag].push(c)
+        end
+      end
+    end
```
Comment (NickPapacostas) [inline app/controllers/metrics_controller.rb:33]: I tested this with a case that had no tags but "food" in the description and it did not appear on the chart. I was able to get it working by removing the blank tags (not sure why those appear)  earlier:

```
  def map_tags(cases)
    #create map, for every tag empty list to collect cases
    tags = Case.tag_options.map{|tag| [tag, []]}.to_h
    cases.each do |c|
      # if the case has no tags, try to read it out of the description
      c_tags = c.tags.reject(&:blank?).empty? ? map_description(c.description) : c.tags.reject(&:blank?)
      c_tags.each do |tag|
        tags[tag].push(c)
      end
    end

    tags
  end
```

File: db/schema.rb [hunk 3]
```diff
@@ -10,8 +10,11 @@
 #
 # It's strongly recommended that you check this file into your version control system.
 
-ActiveRecord::Schema[7.1].define(version: 2024_04_05_151111) do
+ActiveRecord::Schema[7.1].define(version: 2024_04_07_162704) do
+  create_schema "heroku_ext"
+
   # These are extensions that must be enabled in order to support this database
+  enable_extension "pg_stat_statements"
```
Comment (NickPapacostas) [inline db/schema.rb:17]: I don't mind but why do we need this?

**Review by NickPapacostas — COMMENTED** [review]
Nice! just a few comments

File: db/schema.rb [hunk 4]
```diff
@@ -127,20 +119,6 @@
...
-    t.datetime "created_at", precision: nil, null: false
-    t.datetime "updated_at", precision: nil, null: false
-  end
-
-  create_table "linked_documents", force: :cascade do |t|
-    t.bigint "case_id", null: false
-    t.string "title", null: false
-    t.string "url", null: false
-    t.index ["case_id"], name: "index_linked_documents_on_case_id"
-  end
```
[the remark this answers is not in the document]
Comment (melomys) [reply db/schema.rb:142]: those changes apply, when i call db:migrate

**Review by melomys — COMMENTED** [review]
i included your suggestions for the changes in metrics_controller.rb the changes in schema.rb apply, when i call db:migrate, should i not call it, before commiting?

**Comment by NickPapacostas** [comment]
@melomys to be honest I don't remember exactly how it works, maybe those were just never used, I will test a bit after work today

**Pull request merged**